### PR TITLE
Submission Create

### PR DESCRIPTION
**Pull request by Kourchenko**
* Added Create function
* Added Create Unit Tests (Success, Invalid Params)
* Cleanup file structure and added helper functions, helper errors.

### CODE REVIEW

File: src/dal/helper.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    throw errInvalidArgs;
+  }
+
+  // No negative numbers
+  if (accuracy < 0 ||
+    actual_down_speed < 0 ||
+    actual_upload_speed < 0 ||
+    provider_down_speed < 0 ||
+    ping < 0 ||
+    rating < 1) {
```
Comment (Kourchenko) [inline src/dal/helper.ts]: No negative except for rating, which should never really be 0.

File: test/dal/submissions.spec.ts [hunk 2]
```diff
@@ -38,5 +59,77 @@ describe('Submissions DAL', () => {
...
+      sandbox.stub(pool, 'query').callsFake(async () => {
+        return Promise.resolve([createSuccessObj]);
+      });
+      const response = await create(createSuccessObj);
+      expect(response).to.equal(createSuccessObj);
+    });
+
+    it('should handle invalid latitude, longitude values', async () => {
+      sandbox.stub(pool, 'query').callsFake(async () => {
+        return Promise.reject(errInvalidArgs);
```
Comment (ryanrolds) [inline test/dal/submissions.spec.ts]: What returns errInvalidArgs? pool.query or the submissions.create? We also don't need to test individual validation cases in the DAL logic, we can have tests for verifySubmission that ensure it's throwing errors when it should. The submisisons.create logic really only needs one test to confirm an error throw by verifySubmission handled as we expect.

File: test/dal/submissions.spec.ts [hunk 3]
```diff
@@ -38,5 +59,76 @@ describe('Submissions DAL', () => {
...
+      invalidSubmission.ping = -1;
+      invalidSubmission.rating = 0;
+      invalidSubmission.testing_for = 'a'.repeat(256);
+      invalidSubmission.address = 'a'.repeat(256);
+      invalidSubmission.provider = 'a'.repeat(256);
+      invalidSubmission.connected_with = 'a'.repeat(256);
+      invalidSubmission.monthly_price = 'a'.repeat(256);
+      invalidSubmission.hostname = 'a-z!?';
+
+      create(invalidSubmission).catch((err) => {
```
Comment (ryanrolds) [inline test/dal/submissions.spec.ts]: Needs an `await`.

File: test/dal/submissions.spec.ts [hunk 4]
```diff
@@ -38,5 +59,76 @@ describe('Submissions DAL', () => {
...
+    });
+
+    it('should handle query connection error', async () => {
+      sandbox.stub(pool, 'query').callsFake(async () => {
+        return errConnectionRefused;
+      });
+
+      const {host, port} = getConnectionDetails();
+      const errMessage = 'connect ECONNREFUSED ' + host + ':' + port;
+      create(createSuccessObj).catch((err) => {
```
Comment (ryanrolds) [inline test/dal/submissions.spec.ts]: Needs an `await`.

File: test/dal/submissions.spec.ts [hunk 5]
```diff
@@ -38,5 +59,76 @@ describe('Submissions DAL', () => {
...
+
+      create(invalidSubmission).catch((err) => {
+        expect(err).to.equal(errInvalidArgs);
+        expect(err.message).to.equal('Invalid args');
+      });
+    });
+
+    it('should handle query connection error', async () => {
+      sandbox.stub(pool, 'query').callsFake(async () => {
+        return errConnectionRefused;
```
Comment (ryanrolds) [inline test/dal/submissions.spec.ts]: Should return a `Promise.reject(....)`.

File: src/dal/submissions.ts [hunk 6]
```diff
@@ -1,9 +1,42 @@
...
+    'provider_down_speed, rating, ping, hostname) ' +
+    'VALUES (? ,? ,? ,? ,? ,? ,? ,? ,? ,? ,? ,? ,? ,? ,?) ';
+
+  const result = await pool.query(
+      query,
+      [latitude, longitude, accuracy, actual_down_speed,
+        actual_upload_speed, testing_for, address, zip_code,
+        provider, connected_with, monthly_price,
+        provider_down_speed, rating, ping, hostname],
+      (err) => {
```
Comment (ryanrolds) [inline src/dal/submissions.ts]: I don't think function does anything, remove it? It's a callback pattern and we are using promises with the await.

File: package.json [hunk 7]
```diff
@@ -24,6 +25,7 @@
   },
   "homepage": "https://github.com/Hack4Eugene/speedupamerica-frontend#readme",
   "devDependencies": {
+    "@hapi/joi": "^15.1.0",
```
[the remark this answers is not in the document]
Comment (Kourchenko) [reply package.json]: Will move to normal dependency.

File: src/common/errors.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+const errInvalidArgs = new Error('Invalid args');
+const errMissingConnDetails = new Error('database connection details missing');
+const errSubmissionCreate = new Error('Submission Create (Connection)');
+const errConnectionRefused = new Error('ECONNREFUSED');
+
+export {
+  errMissingConnDetails,
+  errInvalidArgs,
+  errSubmissionCreate,
```
[the remark this answers is not in the document]
Comment (Kourchenko) [reply src/common/errors.ts]: Yep yep! They're used in our tests and in our 'dal' functions.

**Pull request merged**